internal/configs/node/config.go

```diff
@@ -251,37 +252,45 @@ func SetShardingSchedule(schedule shardingconfig.Schedule) {
 	}
 }
 
-// ShardIDFromConsensusKey returns the shard ID statically determined from the
-// consensus key.
-func (conf *ConfigType) ShardIDFromConsensusKey() (uint32, error) {
+// ShardIDFromKey returns the shard ID statically determined from the input key
+func (conf *ConfigType) ShardIDFromKey(key *bls.PublicKey) (uint32, error) {
 	var pubKey shard.BLSPublicKey
-	// all keys belong to same shard
-	if err := pubKey.FromLibBLSPublicKey(conf.ConsensusPubKey.PublicKey[0]); err != nil {
+	if err := pubKey.FromLibBLSPublicKey(key); err != nil {
 		return 0, errors.Wrapf(err,
 			"cannot convert libbls public key %s to internal form",
-			conf.ConsensusPubKey.SerializeToHexStr())
+			key.SerializeToHexStr())
 	}
 	epoch := conf.networkType.ChainConfig().StakingEpoch
 	numShards := conf.shardingSchedule.InstanceForEpoch(epoch).NumShards()
 	shardID := new(big.Int).Mod(pubKey.Big(), big.NewInt(int64(numShards)))
 	return uint32(shardID.Uint64()), nil
 }
 
+// ShardIDFromConsensusKey returns the shard ID statically determined from the
+// consensus key.
+func (conf *ConfigType) ShardIDFromConsensusKey() (uint32, error) {
+	return conf.ShardIDFromKey(conf.ConsensusPubKey.PublicKey[0])
+}
+
 // ValidateConsensusKeysForSameShard checks if all consensus public keys belong to the same shard
 func (conf *ConfigType) ValidateConsensusKeysForSameShard(pubkeys []*bls.PublicKey, sID uint32) error {
-	var pubKey shard.BLSPublicKey
+	keyShardStrs := []string{}
+	isSameShard := true
 	for _, key := range pubkeys {
-		if err := pubKey.FromLibBLSPublicKey(key); err != nil {
-			return errors.Wrapf(err,
-				"cannot convert libbls public key %s to internal form",
-				key.SerializeToHexStr())
+		shardID, err := conf.ShardIDFromKey(key)
+		if err != nil {
+			return err
 		}
-		epoch := conf.networkType.ChainConfig().StakingEpoch
-		numShards := conf.shardingSchedule.InstanceForEpoch(epoch).NumShards()
-		shardID := new(big.Int).Mod(pubKey.Big(), big.NewInt(int64(numShards)))
-		if uint32(shardID.Uint64()) != sID {
-			return errors.New("bls keys do not belong to the same shard")
+		if shardID != sID {
+			isSameShard = false
 		}
+		keyShardStrs = append(
+			keyShardStrs,
+			fmt.Sprintf("key: %s, shard id: %d", key.SerializeToHexStr(), shardID),
+		)
+	}
+	if !isSameShard {
+		return fmt.Errorf("bls keys do not belong to same shard\n%s", strings.Join(keyShardStrs, "\n"))
```

Comment:
lets get rid of all `fmt.Errof` and only use `pkg/error`, therefore `errors.Errof`